core/store/src/trie/prefetching_trie_storage.rs

```diff
@@ -0,0 +1,399 @@
+    /// multiple times.
+    pub(crate) prefetching: PrefetchStagingArea,
+}
+
+/// Staging area for in-flight prefetch requests and a buffer for prefetched data.
+///
+/// Before starting a pre-fetch, a slot is reserved for it. Once the data is
+/// here, it will be put in that slot. The parent `TrieCachingStorage` needs
+/// to take it out and move it to the shard cache.
+///
+/// A shared staging area is the interface between `TrieCachingStorage` and
+/// `TriePrefetchingStorage`. The parent simply checks the staging area before
+/// going to the DB. Otherwise, no communication between the two is necessary.
+///
+/// This design also ensures the shard cache works exactly the same with or
+/// without the prefetcher, because the order in which it sees accesses is
+/// independent of the prefetcher.
+#[derive(Default, Clone)]
+pub(crate) struct PrefetchStagingArea(Arc<Mutex<InnerPrefetchStagingArea>>);
+
+#[derive(Default)]
+struct InnerPrefetchStagingArea {
+    slots: HashMap<CryptoHash, PrefetchSlot>,
+    size_bytes: usize,
+}
+
+/// Result when atomically accessing the prefetch staging area.
+pub(crate) enum PrefetcherResult {
+    SlotReserved,
+    Pending,
+    Prefetched(Arc<[u8]>),
+    MemoryLimitReached,
+}
+
+/// Type used interanlly in the staging area to keep track of requests.
+#[derive(Clone, Debug)]
+enum PrefetchSlot {
+    PendingPrefetch,
+    PendingFetch,
+    Done(Arc<[u8]>),
+}
+
+impl TrieStorage for TriePrefetchingStorage {
+    // Note: This is the tricky bit of the implementation.
+    // We have to retrieve data only once in many threads, so all IO threads
+    // have to go though the staging area and check for inflight requests.
+    // The shard cache mutex plus the prefetch staging area mutex are used for
+    // that in combination. Let's call the first lock S and the second P.
+    // The rules for S and P are:
+    // 1. To avoid deadlocks, S must always be requested before P, if they are
+    //    held at the same time.
+    // 2. When looking up if something is already in the shard cache, S must not
+    //    be released until the staging area is updated by the current thread.
+    //    Otherwise, there will be race conditions that could lead to multiple
+    //    threads looking up the same value from DB.
+    // 3. IO threads should release S and P as soon as possible, as they can
+    //    block the main thread otherwise.
+    fn retrieve_raw_bytes(&self, hash: &CryptoHash) -> Result<Arc<[u8]>, StorageError> {
+        // Try to get value from shard cache containing most recently touched nodes.
+        let mut shard_cache_guard = self.shard_cache.0.lock().expect(POISONED_LOCK_ERR);
+        if let Some(val) = shard_cache_guard.get(hash) {
+            return Ok(val);
+        }
+
+        // If data is already being prefetched, wait for that instead of sending a new request.
+        let prefetch_state = PrefetchStagingArea::get_and_set_if_empty(
```

Comment:
nit: why not: self.prefetching.get_and_set_if_empty(...) ?

Reply:
leftover from previous design, thanks!